File: fortran code/dsa.f90
program DSAimpl
implicit none
integer, parameter :: Nx=200,Np=100, Nt=300000
real*8, parameter ::  a=10000.0D0, b=10000.0D0, Pmin=0.01D0, Pmax=100000.0D0
real*8  gn(Nx,Np-1), g(Nx,Np-1), x(1:Nx)
real*8 dt, h1, h2, ymin, dy, y
integer i,k,j,iter


ymin=dlog(Pmin)
dy=dlog(Pmax/Pmin)/Np

!создание сетки по х
h1=0.5D0*Nx/dlog(1.0D0+a)
h2=0.5D0*Nx/dlog(1.0D0+b)
do i=1,Nx/2
	x(i)=1.0D0-exp(-(1.0D0*i-0.5D0*Nx)/h1)
end do
do i=(Nx/2+1),Nx
	x(i)=exp((1.0D0*i-0.5D0*Nx)/h1)-1.0D0
end do
!создание сетки по х

!do i=1,Nx
!	x(i)=-a+(b+a)*i/Nx
!end do


dt=0.001D0
write(*,*) dt, x(Nx/2)-x(Nx/2-1), x(Nx/2+1)-x(Nx/2)
gn=0.0D0
open(8,file='data/gfull.dat')
	do i=1, Nx
		do k=1, Np-1
			read(8,*) gn(i,k)
		end do
	end do
close(UNIT=8)
!gn=0.0D0
g=0.0D0
iter=0
do j=1,Nt
	iter=iter+1
	call solver(a,ymin,x,dy,dt,Nx,Np,gn,g)
	gn(:,:)=g(:,:) 
	write(*,*) iter	
end do


	open(10,file='data/fp4.dat')
	do k=1, Np-1
		y=ymin+k*dy
		write(10,*) k, exp(y), g(Nx/2,k)*exp(y)
	end do
	close(UNIT=10)

	open(8,file='data/Nx.dat')
		write(8,*) Nx
	close(UNIT=8)

	open(12,file='data/Np.dat')
		write(12,*) Np
	close(UNIT=12)

	open(14,file='data/gfull.dat')
	do i=1, Nx
		do k=1, Np-1
			write(14,*) g(i,k)
		end do
	end do
	close(UNIT=14)

	open(18,file='data/x.dat')
	do i=1, Nx
		write(18,*) x(i)
	end do
	close(UNIT=18)


end

File: branches/v3/dsa/dsa.cpp
#include "math.h"
#include "stdio.h"
#include "solver.h"

int main(){

	int Nx=200, Np=100, Nt = 300000;
	double   a=10000, b=10000, Pmin=0.01, Pmax=100000;
	double** gn = new double*[Nx];
	double** g = new double*[Nx];
	for(int i = 0; i < Nx; ++i){
		gn[i] = new double[Np-1];
		g[i] = new double[Np-1];
	}
	double* x = new double[Nx];
	double dt, h1, h2, ymin, dy, y;


	ymin=log(Pmin);
	dy=log(Pmax/Pmin)/Np;

	//создание сетки по х
	h1=0.5*Nx/log(1.0+a);
	h2=0.5*Nx/log(1.0+b);
	for(int i=0; i < Nx/2; ++ i){ 
		x[i] = 1 - exp(-(1.0*(i+1)-0.5*Nx)/h1);
	}
	for(int i=Nx/2; i < Nx; ++i){
		x[i] = exp((1.0*(i+1)-0.5*Nx)/h1)-1.0;
	}

	dt=0.001;
	printf("%lf %g %g\n",dt, x[Nx/2-1]-x[Nx/2-2], x[Nx/2]-x[Nx/2-1]);

	for(int i = 0; i < Nx; ++i){
		for(int j = 0; j < Np-1; ++j){
			gn[i][j] = 0;
			g[i][j] = 0;
		}
	}
	int iter = 0;
	
	for(iter = 1; iter <= Nt; ++iter){
		iter = iter + 1;
		solver(a, ymin, x, dy, dt, Nx, Np, gn, g);
		for(int i = 0; i < Nx; ++i){
			for(int j = 0; j <Np; ++j){
				gn[i][j] = g[i][j];
			}
		}
		printf("%d\n",iter);
	}


	FILE* fp4 = fopen("data/fp4.dat","w");
	for(int k = 0; k < Np-1; ++k){
		y=ymin+k*dy;
		fprintf(fp4,"%d %g %g\n",k, exp(y), g[Nx/2-1][k]*exp(y));
	}
	fclose(fp4);

	FILE* NXFile = fopen("data/Nx.dat","w");
	fprintf(NXFile, "%d", Nx);
	fclose(NXFile);

	FILE* NPFile = fopen("data/Np.dat","w");
	fprintf(NPFile, "%d", Np);
	fclose(NPFile);

	FILE* gfull = fopen("data/gfull.dat","w");
	for(int i = 0; i < Nx; ++i){
		for(int k = 0; k < Np-1; ++k){
			fprintf(gfull, "%lf\n", g[i][k]);
		}
	}
	fclose(gfull);

	FILE* xfile = fopen("data/x.dat","w");
	for(int i = 0; i < Nx; ++i){
		fprintf(xfile, "%lf\n", x[i]);
	}
	fclose(xfile);
}